### PR TITLE
feat: create a new SDK for React Native

### PR DESCRIPTION
**Pull request by omridan159**
## Explanation

This PR introduces the new `sdk-react-native`, a wrapper for the [MetaMask iOS SDK](https://github.com/MetaMask/metamask-ios-sdk) and the [MetaMask Android SDK](https://github.com/MetaMask/metamask-android-sdk). It exposes these SDKs in JavaScript, enabling seamless MetaMask integration into React Native applications. The PR includes the SDK implementation and a comprehensive example demonstrating its usage.
This update simplifies the integration process for developers building React Native dApps with MetaMask support.

https://github.com/MetaMask/metamask-sdk/assets/61094771/9a079056-6ddb-4880-838a-ac5b530ea4db

https://github.com/MetaMask/metamask-sdk/assets/61094771/a165dde6-500b-4622-860f-0dbda3bd03c1

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/examples/reactNativeSdkDemo/package.json [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    "ios": "react-native run-ios --reset-cache --port=4095 --simulator 'iPhone 13 Pro'",
+    "ios:device": "react-native run-ios --port=4095 --device",
+    "start": "react-native start --port=4095",
+    "lint": "yarn lint:eslint",
+    "lint:eslint": "eslint . --cache \"src/**/*.{js,ts,tsx}\"",
+    "test": "jest",
+    "allow-scripts": ""
+  },
+  "dependencies": {
+    "@metamask/sdk-react-native": "file:/Users/omri/Projects/metamask-sdk/packages/sdk-react-native",
```
Comment (omridan159) [inline packages/examples/reactNativeSdkDemo/package.json:16]:
```suggestion
    "@metamask/sdk-react-native": "0.1.0",
```

File: packages/sdk-react-native/README.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import { MetaMaskProvider } from '@metamask/sdk-react-native';
+
+const sdkOptions = {
+  dappMetadata: {
+    name: 'Demo React Native App',
+    url: 'https://yourdapp.com',
+    iconUrl: 'https://yourdapp.com/icon.png',
+    scheme: 'yourappscheme',
+  },
+  infuraAPIKey: 'YOUR_INFURA_API_KEY',
```
Comment (andreahaku) [inline packages/sdk-react-native/README.md:62]: we should add a remark saying that this is optional

File: packages/sdk-react-native/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    </View>
+  );
+};
+
+export default App;
+```
+
+### SDK Options
+
+| Option                 | Type   | Description                             |
```
Comment (andreahaku) [inline packages/sdk-react-native/README.md:110]: add a column explaining what's mandatory and what's optional

File: packages/sdk-react-native/package.json [hunk 4]
```diff
@@ -0,0 +1,107 @@
+{
+  "name": "@metamask/sdk-react-native",
+  "title": "MetaMask - ReactNative SDK",
+  "version": "0.1.0",
+  "description": "TODO",
```
Comment (andreahaku) [inline packages/sdk-react-native/package.json:5]: please add a description

File: packages/sdk-react-native/src/NativePackageMethods.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+import { MetaMaskSDKOptions } from './MetaMaskProvider';
+
+const { MetaMaskReactNativeSdk } = NativeModules;
+
+interface MetaMaskSDKNativeModuleOptions {
+  dappName: string;
+  dappUrl: string;
+  dappIconUrl: string;
+  dappScheme: string;
+  infuraAPIKey: string;
```
Comment (andreahaku) [inline packages/sdk-react-native/src/NativePackageMethods.ts:11]: this should be optional

File: packages/sdk-react-native/src/NativePackageMethods.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+ *
+ * @param sdkOptions - The options for initializing the SDK.
+ */
+export function initializeSDK(sdkOptions: MetaMaskSDKOptions) {
+  const options: MetaMaskSDKNativeModuleOptions = {
+    dappName: sdkOptions.dappMetadata.name,
+    dappUrl: sdkOptions.dappMetadata.url,
+    dappIconUrl: sdkOptions.dappMetadata.iconUrl,
+    dappScheme: sdkOptions.dappMetadata.scheme,
+    infuraAPIKey: sdkOptions.infuraAPIKey,
```
Comment (andreahaku) [inline packages/sdk-react-native/src/NativePackageMethods.ts:136]: optional

File: packages/sdk-react-native/src/SDKConfigProvider.tsx
Comment (andreahaku) [inline packages/sdk-react-native/src/SDKConfigProvider.tsx:1]: make sure that everything related to `InfuraKey` is handled optionally

**Review by andreahaku — COMMENTED** [review]
some comments. thanks

File: packages/sdk-react-native/src/SDKConfigProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+      setTimeout(() => {
+        if (
+          typeof window !== 'undefined' &&
+          // eslint-disable-next-line no-undef
+          typeof window.location !== 'undefined' &&
+          // eslint-disable-next-line no-undef
+          typeof window.location.reload !== 'undefined'
+        ) {
+          // Reload window with changes
+          // eslint-disable-next-line no-undef
```
Comment (abretonc7s) [inline packages/sdk-react-native/src/SDKConfigProvider.tsx:141]: linting disabled

**Review by abretonc7s — CHANGES_REQUESTED** [review]
see comments

File: packages/sdk-react-native/README.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+import { MetaMaskProvider } from '@metamask/sdk-react-native';
+
+const sdkOptions = {
+  dappMetadata: {
+    name: 'Demo React Native App',
+    url: 'https://yourdapp.com',
+    iconUrl: 'https://yourdapp.com/icon.png',
+    scheme: 'yourappscheme',
+  },
+  infuraAPIKey: 'YOUR_INFURA_API_KEY', // Optional
```
Comment (abretonc7s) [inline packages/sdk-react-native/README.md:69]: I would change the comment to say "Optional but highly recommended for a better user experience"

**Review by abretonc7s — APPROVED** [review]
Just tiny comment to add context on infura key but LGTM

**Review by andreahaku — DISMISSED** [review]
LGTM

**Review by andreahaku — DISMISSED** [review]
LGTM

**Review by andreahaku — DISMISSED** [review]
LGTM

**Pull request merged**